Source/Core/AudioCommon/WASAPIStream.cpp

```diff
@@ -82,134 +79,115 @@ static bool HandleWinAPI(std::string message, HRESULT result)
 
 std::vector<std::string> WASAPIStream::GetAvailableDevices()
 {
-  CoInitialize(nullptr);
+  HRESULT result = CoInitializeEx(nullptr, COINIT_MULTITHREADED);
+  // It's fine if CoInitializeEx fails with RPC_E_CHANGED_MODE
+  if (result != RPC_E_CHANGED_MODE && !HandleWinAPI("Failed to call CoInitialize", result))
+    return {};
+
+  wil::unique_couninitialize_call cleanup;
+  if (FAILED(result))
+    cleanup.release();  // Only call CoUninitialize on success
 
-  IMMDeviceEnumerator* enumerator = nullptr;
+  ComPtr<IMMDeviceEnumerator> enumerator;
 
-  HRESULT result =
-      CoCreateInstance(__uuidof(MMDeviceEnumerator), nullptr, CLSCTX_INPROC_SERVER,
-                       __uuidof(IMMDeviceEnumerator), reinterpret_cast<LPVOID*>(&enumerator));
+  result = CoCreateInstance(__uuidof(MMDeviceEnumerator), nullptr, CLSCTX_INPROC_SERVER,
+                            IID_PPV_ARGS(enumerator.GetAddressOf()));
 
   if (!HandleWinAPI("Failed to create MMDeviceEnumerator", result))
     return {};
 
-  std::vector<std::string> device_names;
-  IMMDeviceCollection* devices;
-  result = enumerator->EnumAudioEndpoints(eRender, DEVICE_STATE_ACTIVE, &devices);
+  ComPtr<IMMDeviceCollection> devices;
+  result = enumerator->EnumAudioEndpoints(eRender, DEVICE_STATE_ACTIVE, devices.GetAddressOf());
 
   if (!HandleWinAPI("Failed to get available devices", result))
-  {
-    CoUninitialize();
     return {};
-  }
 
   UINT count;
   devices->GetCount(&count);
 
+  std::vector<std::string> device_names;
+  device_names.reserve(count);
+
   for (u32 i = 0; i < count; i++)
   {
-    IMMDevice* device;
-    devices->Item(i, &device);
+    ComPtr<IMMDevice> device;
+    devices->Item(i, device.GetAddressOf());
     if (!HandleWinAPI("Failed to get device " + std::to_string(i), result))
       continue;
 
-    LPWSTR device_id;
-    device->GetId(&device_id);
-
-    IPropertyStore* device_properties;
+    ComPtr<IPropertyStore> device_properties;
 
-    result = device->OpenPropertyStore(STGM_READ, &device_properties);
+    result = device->OpenPropertyStore(STGM_READ, device_properties.GetAddressOf());
 
     if (!HandleWinAPI("Failed to initialize IPropertyStore", result))
       continue;
 
-    PROPVARIANT device_name;
-    PropVariantInit(&device_name);
-
-    device_properties->GetValue(PKEY_Device_FriendlyName, &device_name);
+    wil::unique_prop_variant device_name;
+    device_properties->GetValue(PKEY_Device_FriendlyName, device_name.addressof());
 
     device_names.push_back(TStrToUTF8(device_name.pwszVal));
-
-    PropVariantClear(&device_name);
   }
 
-  devices->Release();
-  enumerator->Release();
-
-  CoUninitialize();
-
   return device_names;
 }
 
-IMMDevice* WASAPIStream::GetDeviceByName(std::string name)
+ComPtr<IMMDevice> WASAPIStream::GetDeviceByName(std::string name)
 {
-  CoInitialize(nullptr);
+  HRESULT result = CoInitializeEx(nullptr, COINIT_MULTITHREADED);
```

Comment:
This in particular is also done in the constructor, is there a reason why you do it again?

Reply:
No, no reason here, I guess it's a result of copypasting things around from `GetAvailableDevices`. Although I'll oppose removing it, since `GetDeviceByName` is a static public function, so in principle there is nothing preventing it from being called outside the constructor/init logic - unless we want to make it private instead.

Reply:
I completely overlooked they were static, in that sense it makes perfect sense.